examples/mongocxx/instance_management.cpp

```diff
@@ -0,0 +1,110 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstdlib>
+#include <memory>
+#include <mutex>
+
+#include <bsoncxx/stdx/make_unique.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <bsoncxx/stdx/string_view.hpp>
+
+#include <mongocxx/instance.hpp>
+#include <mongocxx/logger.hpp>
+#include <mongocxx/pool.hpp>
+#include <mongocxx/stdx.hpp>
+#include <mongocxx/uri.hpp>
+
+namespace {
+
+// This example demonstrates how one might keep a heap allocated mongocxx::instance and
+// mongocxx::pool object associated in a way that allows dynamic configuration. Most of the examples
+// simply create a stack allocated mongocxx::instance object, which is fine for small examples, but
+// real programs will typically require shared access to a commonly configured instance and
+// connection pool across different translation units and components. By providing a singleton which
+// owns the instance and pool objects, we can defer configuration until we are ready to use MongoDB,
+// and share the instance and pool objects between multiple functions.
+
+class mongo_access {
+public:
+
+    static mongo_access& instance() {
+        std::lock_guard<std::mutex> lock(_singleton_mutex);
+        if (!_singleton)
+            _singleton.reset(new mongo_access);
+        return *_singleton;
+    }
```

Comment:
We could easily enough re-write it to use the Meyer's singleton via magic-statics, if you think that is clearer